README.md
# DataSift Push Server for HTTP

This library is a Node.js implementation for DataSift's HTTP connector as described in their [HTTP docs](http://dev.datasift.com/docs/push/connectors/http). It should be used as a public endpoint for Push delivery.

## Usage overview

The following topics are covered:

* Installation
* Configuration
* Events
* FAQ
* Development & testing

## Installation

This library is not a stand-alone server but should be used as a dependency in an app that is able to handle incoming data.

Add `datasift-push` to your package.json dependencies and run the following command:

    $ npm install
    
Include it in your app:

    var DataSift = require('datasift-push');
    var server = new DataSift({ port: 8080 });
    server.on('data', datahandler);
    
## Configuration

#### `hostname` (String, optional)

Limit incoming requests to a certain hostname

#### `port` (Number, optional)

Listen for incoming requests on a certain port. Defaults to 80.

#### `path` (String, optional)

Limit incoming requests to a certain path. Defaults to "/".  
A path of "/lorem/ipsum" will allow all requests to "/lorem/ipsum" and "/lorem/ipsum/dolor" but not "/lorem/sit/amet"

#### `format` (String, optional)

Format of incoming data. One of `json_meta`, `json_array` or `json_new_line`. Defaults to json_meta.  
Make sure that push subscriptions are created with the same format. The server will refuse all requests that cannot be parsed.

#### `maxSize` (Number, optional)

Maximum allowed size of request body (in bytes). Defaults to 20971520 (20MB, DataSift max)  
The server will immediately close the connection as soon as the size has been exceeded.

#### `confirmData` (Boolean, optional)

Expect app to confirm all emitted data. Defaults to false.  
When set to true, the server expects your app to confirm that it has processed incoming data. The `data` event will have a third parameter with a "done" callback. Call this to confirm that the incoming data has been saved. If your app does not respond within 10 seconds, the server will automatically respond with a HTTP 500 message. As a result, DataSift will try to resend the data after a certain interval.

#### `username` and `password` (String, optional)

Requires incoming requests to be authorized. Defaults to undefined.

### Example

    var server = new DataSift({
        port: 8080,
        path: '/lorem/ipsum/',
        username: 'john',
        password: 'loremipsum'
    });

## Events

#### `initialized`

Emitted when server starts listening on the provided port. Arguments: `port`, `hostname`

#### `refused`

Emitted when an incoming request has been refused. Arguments: `reason`

- Could not inflate compressed data
- Request did not contain required authentication headers
- Requested path "..." does not match "..."
- Post data exceeded limit of ... bytes
- Could not parse post data

#### `error` 

Emitted on internal errors. Arguments: `error` (Error instance)

#### `check`

Emitted after handling a "check" request from DataSift. Arguments: none

#### `data`

Emitted after parsing incoming data. Arguments: `data`, `meta`, `done` (optional)

- data: array of interactions and deletions
- meta: object with the request's meta data
    - url (raw url)
    - url_parsed (parsed url)
    - hash (playback id or stream hash)
    - hash_type (either historic or stream)
    - id (subscription id)
- done: function to be called after processing incoming data. Only provided when option `confirmData` is true. Call as `done(false)` to respond with a HTTP 500.

## FAQ

Q: I do not receive any incoming data  
A: Make sure the path, hostname and port are configured correctly and that your server is accessible from the internet.

Q: I keep receiving the same data  
A: Did you enable the `confirmData` option? Make sure you call the `done` function after processing incoming data. Make sure you are able to process incoming data in time.

Q: I am unable to see what's going on  
A: Listen to all described events to find out if there are any errors or refused connections. You might want to use `server.http` for more details debugging.


## Developing

The library is published to NPM and can be installed with the following command:

    $ npm install datasift-push

## Testing

Navigate to this module's repository and make sure you have the development modules installed:

    $ npm install


Run the tests:

    $ npm test



CHANGELOG.md
# Changelog

## 0.2.0 (2013-10-15)

- Documentation in README.md
- Added Mocha tests
- Added meta data to data event arguments
- Minor bugfixes

## 0.1.0 (2013-10-10)

- First release!


test/mocks.js
var events = require('events');
var util = require('util');

// Mock http.createServer
var httpServer = function(){
    events.EventEmitter.apply(this, arguments);
};
util.inherits(httpServer, events.EventEmitter);
httpServer.prototype.listen = function(port, hostname, callback){
    process.nextTick(callback);
};

// Mock incoming data
var interactions = [
    {interaction: 'foo', twitter: 'bar'},
    {interaction: 'lorem', twitter: 'ipsum'},
    {interaction: 'dolor', other: 'amet'}
];
var data = {
    jsonMeta: JSON.stringify({id: 123, interactions: interactions}),
    jsonArray: JSON.stringify({interactions: interactions}),
    jsonNewLine: interactions.map(JSON.stringify).join("\n"),
    check: '{}',
    invalid: 'foo bar'
};

// Export mocks
module.exports = {
    createServer: function(){ return new httpServer(); },
    data: data
};

test/server.js
var http = require('http');
var expect = require('expect.js');
var sinon = require('sinon');
var httpMocks = require('node-mocks-http');
var mocks = require('./mocks');
var Server = require('../lib/server');

describe('Server', function(){

    var spy, s, request, response, _RESPOND_TIME;
    beforeEach(function(){
        spy = sinon.spy();
        s = new Server();
        request = httpMocks.createRequest();
        response = httpMocks.createResponse();
        _RESPOND_TIME = Server.RESPOND_TIME;
    });

    afterEach(function(){
        Server.RESPOND_TIME = _RESPOND_TIME;
    });

    before(function(){
        sinon.stub(http, 'createServer', mocks.createServer);
    });

    after(function(){
        http.createServer.restore();
    });

    describe('Constants', function(){
        it('should have a default hostname', function(){
            expect(Server.DEFAULT_HOSTNAME).to.be(undefined);
        });

        it('should have a default port', function(){
            expect(Server.DEFAULT_PORT).to.be(80);
        });

        it('should have a default path', function(){
            expect(Server.DEFAULT_PATH).to.be('/');
        });

        it('should have a default format', function(){
            expect(Server.DEFAULT_FORMAT).to.be('json_meta');
        });

        it('should have a default confirm_date', function(){
            expect(Server.CONFIRM_DATA).to.be(false);
        });

        it('should have a default max_size', function(){
            expect(Server.MAX_SIZE).to.be(20971520);
        });

        it('should have a default respond_time', function(){
            expect(Server.RESPOND_TIME).to.be(10000);
        });
    });

    describe('Constructor', function(){
        it('should set default options', function(){
            var s = new Server();
            expect(s.hostname).to.be(Server.DEFAULT_HOSTNAME);
            expect(s.port).to.be(Server.DEFAULT_PORT);
            expect(s.path).to.be(Server.DEFAULT_PATH);
            expect(s.format).to.be(Server.DEFAULT_FORMAT);
            expect(s.confirmData).to.be(Server.CONFIRM_DATA);
            expect(s.maxSize).to.be(Server.MAX_SIZE);
        });

        it('should keep track of config options', function(){
            var options = {foo: 'bar'};
            var s = new Server(options);
            expect(s.options).to.eql(options);
        });
    });

    describe('Ready state', function(){
        it('should emit initialized event', function(done){
            var s = new Server({ port: 1234, hostname: 'example.com' });
            s.on('initialized', function(){
                spy.apply(spy, arguments);
                expect(spy.args[0][0]).to.be(1234);
                expect(spy.args[0][1]).to.be('example.com');
                done();
            });
        });
    });

    describe('#handleError', function(){
        it('should emit error', function(){
            var error = new Error('Lorem ipsum');
            s.on('error', spy);
            s.handleError(error);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be(error);
        });
    });

    describe('#handleRequestData', function(){
        it('should emit data event on valid request', function(){
            var rawData = mocks.data.jsonMeta;
            s.on('data', spy);
            s.handleRequestData(request, response, rawData);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be.an('array');
            expect(spy.args[0][0].length).to.be(3);
        });

        it('should not emit data event on invalid request', function(){
            var rawData = mocks.data.invalid;
            s.on('data', spy);
            s.handleRequestData(request, response, rawData);
            expect(spy.callCount).to.be(0);
        });

        it('should not emit data event on check request', function(){
            var rawData = mocks.data.check;
            s.on('data', spy);
            s.handleRequestData(request, response, rawData);
            expect(spy.callCount).to.be(0);
        });
    });

    describe('#respond', function(){
        it('should respond with default values', function(){
            s.respond(response);
            expect(response.statusCode).to.be(200);
            expect(response.getHeader('Content-Type')).to.be('text/plain');
            expect(response._getData()).to.be(JSON.stringify({success: true}));
        });

        it('should respond with custom code', function(){
            s.respond(response, 404);
            expect(response.statusCode).to.be(404);
            expect(response.getHeader('Content-Type')).to.be('text/plain');
            expect(response._getData()).to.be('Not Found');
        });

        it('should respond with custom data', function(){
            s.respond(response, undefined, 'lorem');
            expect(response.statusCode).to.be(200);
            expect(response.getHeader('Content-Type')).to.be('text/plain');
            expect(response._getData()).to.be('lorem');
        });

        it('should respond with custom code and data', function(){
            s.respond(response, 404, 'lorem');
            expect(response.statusCode).to.be(404);
            expect(response.getHeader('Content-Type')).to.be('text/plain');
            expect(response._getData()).to.be('lorem');
        });
    });

    describe('#validateAuthentication', function(){
        it('should emit refused event on missing credentials', function(){
            var s = new Server({username: 'lorem', password: 'ipsum'});
            s.on('refused', spy);
            var result = s.validateAuthentication(request, response);
            expect(response.statusCode).to.be(401);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be('Request did not contain required authentication headers');
            expect(result).to.be(false);
        });

        it('should not emit refused event on optional credentials', function(){
            s.on('refused', spy);
            var result = s.validateAuthentication(request, response);
            expect(spy.callCount).to.be(0);
            expect(result).to.be(true);
        });

        it('should not emit refused event on correct credentials', function(){
            var s = new Server({username: 'lorem', password: 'ipsum'});
            request._setHeadersVariable('authorization', 'Basic ' + new Buffer('lorem:ipsum').toString('base64'));
            s.on('refused', spy);
            var result = s.validateAuthentication(request, response);
            expect(spy.callCount).to.be(0);
            expect(result).to.be(true);
        });
    });

    describe('#validatePath', function(){
        it('should emit refused event on invalid path', function(){
            var s = new Server({path: '/lorem/ipsum'});
            request._setURL('/foo/bar');
            s.on('refused', spy);
            var result = s.validatePath(request, response);
            expect(response.statusCode).to.be(404);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be('Requested path "/foo/bar" does not match "/lorem/ipsum"');
            expect(result).to.be(false);
        });

        it('should not emit refused event on valid path', function(){
            s.on('refused', spy);
            request._setURL('/');
            var result = s.validatePath(request, response);
            expect(spy.callCount).to.be(0);
            expect(result).to.be(true);
        });
    });

    describe('#validateSize', function(){
        it('should emit refused event on invalid size', function(){
            request.connection = {destroy: function(){}}; // mock connection
            var s = new Server({maxSize: 500});
            s.on('refused', spy);
            var result = s.validateSize(request, response, 501);
            expect(response.statusCode).to.be(413);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be('Post data exceeded limit of 500 bytes');
            expect(result).to.be(false);
        });

        it('should not emit refused event on valid size', function(){
            s.on('refused', spy);
            var result = s.validateSize(request, response, 500);
            expect(spy.callCount).to.be(0);
            expect(result).to.be(true);
        });
    });

    describe('#handleCheck', function(){
        it('should emit check event on valid data', function(){
            s.on('check', spy);
            var result = s.handleCheck(request, response, mocks.data.check);
            expect(response.statusCode).to.be(200);
            expect(spy.callCount).to.be(1);
            expect(result).to.be(true);
        });

        it('should not emit check event on invalid data', function(){
            s.on('check', spy);
            var result = s.handleCheck(request, response, mocks.data.invalid);
            expect(spy.callCount).to.be(0);
            expect(result).to.be(false);
        });
    });

    describe('#parse', function(){
        it('should emit refused event on invalid data', function(){
            s.on('refused', spy);
            var result = s.parse(request, response, mocks.data.invalid);
            expect(response.statusCode).to.be(400);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.be('Could not parse post data');
            expect(result).to.be(undefined);
        });

        it('should not emit refused event on valid data', function(){
            s.on('refused', spy);
            var result = s.handleCheck(request, response, mocks.data.jsonMeta);
            expect(spy.callCount).to.be(0);
            expect(result).not.to.be(undefined);
        });

        it('should return parsed data with format json_meta', function(){
            s = new Server({format: 'json_meta'});
            var interactions = JSON.parse(mocks.data.jsonMeta).interactions;
            var result = s.parse(request, response, mocks.data.jsonMeta);
            expect(result).to.eql(interactions);
        });

        it('should return parsed data with format json_array', function(){
            s = new Server({format: 'json_array'});
            var interactions = JSON.parse(mocks.data.jsonArray).interactions;
            var result = s.parse(request, response, mocks.data.jsonArray);
            expect(result).to.eql(interactions);
        });

        it('should return parsed data with format json_new_line', function(){
            s = new Server({format: 'json_new_line'});
            var interactions = mocks.data.jsonNewLine.split("\n").map(JSON.parse);
            var result = s.parse(request, response, mocks.data.jsonNewLine);
            expect(result).to.eql(interactions);
        });

        it('should return undefined with mismatching format', function(){
            var result = s.parse(request, response, mocks.data.jsonNewLine);
            expect(result).to.eql(undefined);
        });
    });

    describe('#emitData', function(){
        var data = [];
        it('should emit data event with data and meta', function(){
            s.on('data', spy);
            s.emitData(request, response, data);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.eql(data);
            expect(spy.args[0][1]).to.be.an('object'); // meta
            expect(spy.args[0][2]).to.be(undefined);
        });

        it('should emit data event with data, meta and done', function(){
            var s = new Server({ confirmData: true });
            s.on('data', spy);
            s.emitData(request, response, data);
            expect(spy.callCount).to.be(1);
            expect(spy.args[0][0]).to.eql(data);
            expect(spy.args[0][1]).to.be.an('object'); // meta
            expect(spy.args[0][2]).to.be.an('function');
        });

        it('should respond with success on default confirmData', function(){
            s.emitData(request, response, data);
            expect(response.statusCode).to.be(200);
        });

        it('should not respond when confirmData is true', function(){
            var s = new Server({ confirmData: true });
            s.emitData(request, response, data);
            expect(response.statusCode).to.be(-1);
        });

        it('should respond after a while with HTTP 500 when confirmData is true', function(done){
            Server.RESPOND_TIME = 1;
            var errorSpy = sinon.spy();
            var s = new Server({ confirmData: true });
            var response = httpMocks.createResponse();
            s.on('error', errorSpy);
            s.emitData(request, response, data);
            setTimeout(function(){
                expect(response.statusCode).to.be(500);
                expect(errorSpy.callCount).to.be(1);
                expect(errorSpy.args[0][0].message).to.be('Callback for data event not called within allowed time');
                done();
            }, 1);
        });
    });
});

lib/server.js
var events = require('events');
var util = require('util');
var http = require('http');
var zlib = require('zlib');
var url = require('url');

/**
 * Initialize class
 */
function Server(){
    events.EventEmitter.call(this);
    this.initialize.apply(this, arguments);
}
util.inherits(Server, events.EventEmitter);

/**
 * Class constants
 */
Server.DEFAULT_HOSTNAME = undefined;
Server.DEFAULT_PORT = 80;
Server.DEFAULT_PATH = '/';
Server.DEFAULT_FORMAT = 'json_meta'; // json_meta, json_array, json_new_line
Server.CONFIRM_DATA = false;
Server.MAX_SIZE = 20971520; // 20MB, DataSift max
Server.RESPOND_TIME = 10000; // 10sec, DataSift maximum respond time

/**
 * Constructor
 *
 * @param {Object} options Configuration options (optional)
 */
Server.prototype.initialize = function(options){

    // Apply options
    if(!options) options = {};
    this.options = options;
    this.hostname = options.hasOwnProperty('hostname')? options.hostname : Server.DEFAULT_HOSTNAME;
    this.port = options.port || Server.DEFAULT_PORT;
    this.path = options.path || Server.DEFAULT_PATH;
    this.format = options.format || Server.DEFAULT_FORMAT;
    this.confirmData = options.hasOwnProperty('confirmData')? options.confirmData : Server.CONFIRM_DATA;
    this.maxSize = options.maxSize || Server.MAX_SIZE;

    // Start server
    this.http = http.createServer(this.handleRequest.bind(this));
    this.http.on('error', this.handleError.bind(this));
    this.http.listen(this.port, this.hostname, function(){
        this.emit('initialized', this.port, this.hostname);
    }.bind(this));
};

/**
 * Handle incoming request
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 */
Server.prototype.handleRequest = function(request, response){
    var buffer = [];
    var size = 0;

    // Validate authentication
    if(!this.validateAuthentication(request, response)){ // sends proper response on fail
        return;
    }

    // Validate path
    if(!this.validatePath(request, response)){ // sends proper response on fail
        return;
    }

    // Handle errors
    request.on('error', this.handleError.bind(this));

    // Handle incoming data chunks
    request.on('data', function(chunk){
        if(request.connection.destroyed) return;

        // Validate data size
        size += chunk.length;
        if(!this.validateSize(request, response, size)){ // sends proper response on fail
            return;
        }

        // Assemble post data
        buffer.push(new Buffer(chunk));
    }.bind(this));

    // Handle finished request
    request.on('end', function(){
        if(request.connection.destroyed) return;
        var bufferData = Buffer.concat(buffer);
        if(request.headers['content-encoding'] === 'gzip'){
            zlib.inflate(bufferData, function(error, data){
                if(error){
                    this.respond(response, 400);
                    this.emit('refused', 'Could not inflate compressed data');
                }else{
                    this.handleRequestData(request, response, data.toString());
                }
            }.bind(this));
        }else{
            this.handleRequestData(request, response, bufferData.toString());
        }
    }.bind(this));
};

/**
 * Emit internal errors
 *
 * @param {Error} e
 */
Server.prototype.handleError = function(e){
    this.emit('error', e);
};

/**
 * Validate and handle request body
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @param {String} rawData
 */
Server.prototype.handleRequestData = function(request, response, rawData){

    // Respond to checks
    if(this.handleCheck(request, response, rawData)){
        return;
    }

    // Parse and validate data
    var data = this.parse(request, response, rawData); // sends proper response on fail
    if(!data) return;

    // Emit data and respond to request
    this.emitData(request, response, data);
};

/**
 * Respond to request
 *
 * @param {http.ServerResponse} response
 * @param {number} code HTTP status code (optional, default 200)
 * @param {String} data Response body (optional)
 */
Server.prototype.respond = function(response, code, data){
    if(!code) code = 200;
    response.writeHead(code, http.STATUS_CODES[code], { 'Content-Type': 'text/plain' });
    if(code === 200 && !data){
        response.end(JSON.stringify({success: true}));
    }else{
        response.end(data || http.STATUS_CODES[code]);
    }
};

/**
 * Handle HTTP authentication
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @return {Boolean} true on success, false on fail
 */
Server.prototype.validateAuthentication = function(request, response){

    // Ok when no authorization required
    if(!(this.options.username && this.options.password)) return true;

    // Validate authorization headers
    var header = request.headers.authorization || '';
    var encrypted = header.split(' ').pop() || '';
    var credentials = new Buffer(encrypted, 'base64').toString().split(':');
    if(credentials[0] === this.options.username && credentials[1] === this.options.password) return true;

    // Respond with 401:unauthorized
    this.respond(response, 401);
    this.emit('refused', 'Request did not contain required authentication headers');
    return false;
};

/**
 * Validate requested path
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @return {Boolean} true on success, false on fail
 */
Server.prototype.validatePath = function(request, response){
    var path = url.parse(request.url, true).pathname;
    if(path.indexOf(this.path) === 0){
        return true;
    }else{
        this.respond(response, 404);
        this.emit('refused', 'Requested path "' + path + '" does not match "' + this.path + '"');
        return false;
    }
};

/**
 * Validate size of post data
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @param {number} size
 * @return {Boolean} true on success, false on fail
 */
Server.prototype.validateSize = function(request, response, size){
    if(size <= this.maxSize){
        return true;
    }else{
        this.respond(response, 413);
        request.connection.destroy();
        this.emit('refused', 'Post data exceeded limit of ' + this.maxSize + ' bytes');
        return false;
    }
};

/**
 * Handle endpoint checks (post body is "{}")
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @param {String} rawData
 * @return {Boolean} true when request was a check, false otherwise
 */
Server.prototype.handleCheck = function(request, response, rawData){
    if(rawData === '{}'){
        this.respond(response, 200);
        this.emit('check');
        return true;
    }else{
        return false;
    }
};

/**
 * Parse post data
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @param {String} rawData
 * @return {Array} List of interactions
 */
Server.prototype.parse = function(request, response, rawData){
    try{
        if(this.format === 'json_meta' || this.format === 'json_array'){
            return JSON.parse(rawData).interactions;
        }else{
            return rawData.trim().split("\n").map(JSON.parse);
        }
    }catch(e){
        this.respond(response, 400);
        this.emit('refused', 'Could not parse post data');
        return undefined;
    }
};

/**
 * Emit parsed data
 *
 * @param {http.IncomingMessage} request
 * @param {http.ServerResponse} response
 * @param {Array} data List of interactions
 */
Server.prototype.emitData = function(request, response, data){

    // Callback for successful handles
    var doneTimeout;
    var done = function(success){
        clearTimeout(doneTimeout);
        this.respond(response, success === false? 500 : 200);
    }.bind(this);

    // Run callback if no response within proper time
    doneTimeout = setTimeout(function(){
        this.emit('error', new Error('Callback for data event not called within allowed time'));
        done(false);
    }.bind(this), Server.RESPOND_TIME);

    // Assemble meta data
    var meta = {
        url: request.url,
        url_parsed: url.parse(request.url, true),
        hash: request.headers['x-datasift-hash'],
        hash_type: request.headers['x-datasift-hash-type'],
        id: request.headers['x-datasift-id']
    };

    // Emit data and execute success callback
    if(this.confirmData){
        this.emit('data', data, meta, done);
    }else{
        this.emit('data', data, meta);
        done();
    }
};

module.exports = Server;
